contracts/VenusChainlinkOracle.sol

```diff
@@ -0,0 +1,106 @@
+pragma solidity 0.5.16;
+
+import "./PriceOracle.sol";
+import "./VBep20.sol";
+import "./BEP20Interface.sol";
+import "./SafeMath.sol";
+import "./AggregatorV2V3Interface.sol";
+
+contract VenusChainlinkOracle is PriceOracle {
+    using SafeMath for uint;
+    uint public constant VAI_VALUE = 1e18;
+    address public admin;
+
+    mapping(address => uint) internal prices;
+    mapping(bytes32 => AggregatorV2V3Interface) internal feeds;
+    event PricePosted(address asset, uint previousPriceMantissa, uint requestedPriceMantissa, uint newPriceMantissa);
+    event NewAdmin(address oldAdmin, address newAdmin);
+    event FeedSet(address feed, string symbol);
+
+    constructor() public {
+        admin = msg.sender;
+    }
+
+    function getUnderlyingPrice(VToken vToken) public view returns (uint) {
+        string memory symbol = vToken.symbol();
+        if (compareStrings(symbol, "vBNB")) {
+            return getChainlinkPrice(getFeed(symbol));
+        } else if (compareStrings(symbol, "VAI")) {
+            return VAI_VALUE;
+        } else if (compareStrings(symbol, "XVS")) {
+            return prices[address(vToken)];
+        } else {
+            return getPrice(vToken);
+        }
+    }
+
+    function getPrice(VToken vToken) internal view returns (uint) {
+        uint price;
+        BEP20Interface token = BEP20Interface(VBep20(address(vToken)).underlying());
+
+        if (prices[address(token)] != 0) {
+            price = prices[address(token)];
+            uint decimalDelta = uint(18).sub(uint(token.decimals()));
+            // Ensure that we don't multiply the result by 0
+            if (decimalDelta > 0) {
+                return price.mul(10**decimalDelta);
+            } else {
+                return price;
+            }
+        } else {
+            return getChainlinkPrice(getFeed(token.symbol()));
+        }
+    }
+
+    function getChainlinkPrice(AggregatorV2V3Interface feed) internal view returns (uint) {
+        // Chainlink USD-denominated feeds store answers at 8 decimals
+        uint decimalDelta = uint(18).sub(feed.decimals());
+        // Ensure that we don't multiply the result by 0
+        if (decimalDelta > 0) {
+            return uint(feed.latestAnswer()).mul(10**decimalDelta);
+        } else {
+            return uint(feed.latestAnswer());
```

Comment:
@thodges-gh thanks for your update so far.
we need update here for different underlying decimals.
our price oracle needs to have decimal of `18 + (18 - underlying token.decimal)`, not just 18 since of calculation on the protocol.
let me give you an example.
```
USDT: 1000000000000000000000000000000
USDC: 1000000000000000000000000000000
BUSD: 1000000000000000000
```
(here lets say USDT, USDC underlying decimal is 6, not 18: our testnet USDT, USDC mock token has 6 decimal)

### You can check original price oracle values here.
https://testnet.bscscan.com/address/0xb7D29B30a04d0030F832Db32038360fD300A1590
vToken: 
- 0xD5C4C2e2facBEB59D0216D0595d63FcDc6F9A1a7(USDC)
- 0xb7526572FFE56AB9D7489838Bf2E18e3323b441A(USDT)
- 0x08e0A5575De71037aE36AbfAfb516595fE68e5e4(BUSD)

Reply:
Good info! This is now updated.